peak/float.py

```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
-                return Float.reinterpret_as_bv(res_f)
-        return Add
-
-    @family_closure
-    def Sub_fc(family: TypeFamily):
-        Float = Float_fc(family)
-        @family.assemble(locals(), globals())
-        class Sub(Peak, BlackBox):
-            def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
-                in0_f = Float.reinterpret_from_bv(in0)
-                in1_f = Float.reinterpret_from_bv(in1)
-                res_f = in0_f - in1_f
-                return Float.reinterpret_as_bv(res_f)
-        return Sub
-
-    @family_closure
-    def Mul_fc(family: TypeFamily):
-        Float = Float_fc(family)
-        @family.assemble(locals(), globals())
-        class Mul(Peak, BlackBox):
-            def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
-                in0_f = Float.reinterpret_from_bv(in0)
-                in1_f = Float.reinterpret_from_bv(in1)
-                res_f = in0_f * in1_f
-                return Float.reinterpret_as_bv(res_f)
-        return Mul
-
-    @family_closure
-    def Div_fc(family: TypeFamily):
-        Float = Float_fc(family)
-        @family.assemble(locals(), globals())
-        class Div(Peak, BlackBox):
-            def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
-                in0_f = Float.reinterpret_from_bv(in0)
-                in1_f = Float.reinterpret_from_bv(in1)
-                res_f = in0_f / in1_f
-                return Float.reinterpret_as_bv(res_f)
-        return Div
-
-    @family_closure
-    def Fma_fc(family: TypeFamily):
-        Float = Float_fc(family)
-        @family.assemble(locals(), globals())
-        class Fma(Peak, BlackBox):
-            def __call__(self, rm: RoundingMode, in0: Data, in1: Data, in2: Data) -> Data:
-                in0_f = Float.reinterpret_from_bv(in0)
-                in1_f = Float.reinterpret_from_bv(in1)
-                in2_f = Float.reinterpret_from_bv(in1)
-                res_f = in0_f * in1_f + in2_f
-                return Float.reinterpret_as_bv(res_f)
-        return Fma
-
-    @family_closure
-    def Sqrt_fc(family: TypeFamily):
-        @family.assemble(locals(), globals())
-        class Sqrt(Peak, BlackBox):
-            def __call__(self, rm: RoundingMode, in_: Data) -> Data:
-                ...
-        return Sqrt
+    def _get_cast(family):
+        if isinstance(family, MagmaFamily):
+            return _cast_magma
+        elif isinstance(family, SMTFamily):
+            return _cast_smt
+        else:
+            return _cast_py
 
+    @lru_cache(None)
+    def gen_binary(op_name):
+        @family_closure
+        def fc(family: TypeFamily):
+            cast = _get_cast
+            @family.assemble(locals(), globals())
+            class Op(Peak, BlackBox):
+                def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
+                    if rm == RoundingMode.RNE:
+                        in0_rne = cast(RM.RNE, in0)
```

Comment:
Basically of this is to cast between hwtypes rounding mode and floatlib rounding mode.